docs/config.md
# Configuration

Kiln configuration can either be done by passing command line arguments (e.g.
`--network babylonnet`) or by putting a config file with the argument contents
into a file in the config directory (e.g. a file `config/network` containing
`babylonnet`).

## loggers

See [logging.md](https://gitlab.com/tezos-kiln/kiln/-/blob/develop/docs/logging.md)

## pg-connection CONNSTRING

Connection string or URI to PostgreSQL database. If blank, use connection string
in 'db' file or create a database there if empty.

## route URL

Root URL for this service as seen by external users. If blank, use contents of
'config/route'.

## email-from EMAIL

Email address to use for 'From' field in email notifications. If blank, use
contents of 'config/email-from'.

## check-for-upgrade BOOL

Enable/disable upgrade checks. If blank, use contents of
'config/check-for-upgrade'. If that is blank, default to enabled.

## upgrade-branch BRANCH

Upstream Git branch to use for checking upgrades. If blank, use contents of
'config/upgrade-branch'. If that is blank, default to 'master'.

## network NETWORK

Name of a network (e.g mainnet, mumbainet), url of network config or network ID
to monitor. If blank, use contents of 'config/network'. If also blank, default to 'mainnet'.

## tzscan-api-uri URL

Custom tzscan API URL.  Default none.

## blockscale-api-uri URL

Custom Blockscale API URL.  Default none.

## nodes URIS

Force the set of monitored nodes to be exactly the given set of
(comma-separated) list of nodes. If given multiple times, the sets will be
unioned. Defaults to off.

## bakers PUBLICKEYHASHES

Force the set of monitored bakers to be exactly the given set of
(comma-separated) list of bakers. If given multiple times, the sets will be
unioned. Defaults to off.

## network-gitlab-project-id PROJECTID

The GitLab project id to query for network updates. Defaults to off.

## kiln-node-rpc-port PORT

The RPC port to use for the kiln node. Defaults to 8733.

## kiln-node-net-port PORT

The port where kiln node accepts peer-to-peer connections. Defaults to 9733.

## kiln-data-dir DIRECTORY

The data directory used by the kiln node and tezos-client. Defaults to
"./.kiln".

## kiln-node-custom-args ARGS

Custom arguments for the Kiln Node.

## kiln-baker-custom-args ARGS

Custom arguments for the Kiln baker daemon.

## check-ledger-connection BOOL

Enable/disable ledger connection checks when Kiln Baker doesn't have rights.
If blank, use contents of 'config/check-ledger-connection'. If that is blank,
default to enabled.

If this option is disabled, the ledger indicator on the header will be shown
only if Kiln Baker has baking rights.

## resolved-alerts-ttl INT

Time in minutes after which resolved alerts are deleted. The value should be between 0 and 10080 (7 days). Defaults to 4320 minutes (3 days).

## dal-node-uri URI

Octez Data Availability Layer node URI. If not provided, a DAL node won't be used.

## binary-paths

Create a file named `binary-paths` in the `config` directory adjacent
to the `backend` executable.

```json
{
    "node-path" : "<tezos-node-path>",
    "client-path" : "<tezos-client-path>",
    "baker-paths" :
    [
        { "proto" : "<protocol-hash>",
          "baker-path" : "<tezos-baker-path>"
        }
    ]
}
```

This configuration specifies the respective locations of these
binaries: tezos-<node,client,baker>. Notice that the protocol
hash of the desired network must also be included.


CHANGELOG.md
# Changelog

## 0.8.1

  * Bump Tezos mainnet node to include fixes to the mempool and the Tezos node
    (f0607f4b1f8c0f2a0a48adb4bc76378406203714)

### Known Issues

  * Kiln's memory usage may spike at the startup for few minutes.
    It is recommended to have swap memory enabled to avoid causing issues, especially if the tezos-node is also running on the same machine.
    Disabling the OS Public Node via the config file could reduce the Kiln's memory consumption further.

## 0.8.0

  * Updated tezos node dependencies to support the Carthage protocol and the new, more compact storage backend (irmin-pack).

  * Updated the Kiln Node management code to require the new tezos node, with a prompt to delete the old node and to initialize a new one if the existing data store is not at least version 0.0.4.

  * Tweaked the protocol constant parsing to support Carthage.

  * Support for the Carthagenet test network. (Babylonnet is still available for the time being.)

## 0.7.4

  * Ledger connectivity check is now disabled by default.
    To enable the check use the 'ledger-check-delay' option.

  * Bug Fixes
    * Fix the export logs functionality for Debian.

## 0.7.3

  * Ledger connectivity icon and alert
    Kiln now displays an icon on the top bar to indicate if the Ledger is connected/disconnected,
    and also gives an alert to the user if the Ledger device gets disconnected.

  * Support for baking with the new ed25519 derivation scheme (bip25519)

  * Support for babylonnet. (alphanet has been removed)

  * Snapshot import can be cancelled.

  * Improved log messages.
    Kiln now gives better log messages when it is not able to obtain required information from available nodes.

  * Support building Kiln on MacOS from source code using nix

  * "New Tezos software version is available" notification logic has been improved.
    Kiln now detects the version of each monitored node separately, and notify on the node tile UI if the node is running the software different from the latest version.
    This feature can also be disabled / configured via command line options.

  * TzScan public node has been removed.

  * Block hashes on node tile now links to tzstats.com (previously tzscan.com was used).

  * Bug Fixes
    * In some cases Kiln would continue to show "Gathering baking data", even though it has obtained all the available data (of preserved cycles) from the nodes.
      This has been fixed.
    * Fixed baker alert logic to not show the alert if the network is changed.
    * Babylon specific
      * Block's operation parsing now works if the entrypoint is an empty string.
      * Double endorsement detection logic has been fixed.

  * UI improvements
    * When adding a new baker, the Public Key Hash's curve and derivation info is shown as a hover tool tip.
    * Hovering over 'Network' in Kiln's header shows the current protocol.

  * Code improvements
    * The underlying app frameworks used by Kiln (Rhyolite, Obelisk and Reflex) have been upgraded.
    * The core logic of Tezos block data and node RPC communications, written in Haskell, has been moved to a separate project (https://gitlab.com/obsidian.systems/tezos-bake-monitor-lib)
      This will allow anyone to easily make use of this code to create other projects for Tezos.

### Known Issues
  * The copy button on the 'Network' -> 'Current Protocol' pop up might fail to copy the protocol hash to the clipboard.
  * On cancelling snapshot import the UI might display "Snapshot import failed" error.

## 0.7.2

  * Include the tezos-node containing: Mempool: remove parsing exceptions (a7d357bb)

    https://gitlab.com/tezos/tezos/commit/a7d357bb290eab43c92ec2245811e9af959d1d4c

### Known Issues
    * Upgrading Kiln from v0.6.* on Ubuntu/Debian overwrites the /etc/kiln/args file
    * Kiln will not work on zeronet.

## 0.7.1

  * Support for Babylon protocol (005_PsBabyM1)
    * This release has a fix to force-run the PsBabyM1 baker/endorser binaries.
      This is required to avoid missing the bake/endorse opportunity for the first block of the new protocol.

    * With 0.7.0 users can miss bake/endorse opportunities for first couple of blocks in the new protocol.
      It is recommended to upgrade to 0.7.1 before the start of Babylon protocol.

  * This removes the BABY5H baker/endorser, so kiln will not work on zeronet.

### Known Issues
    * Upgrading Kiln on Ubuntu/Debian overwrites the /etc/kiln/args file

## 0.7.0

  * Support for Babylon protocol (005_PsBabyM1 and 005_PsBABY5H)
    * Kiln now has tezos binaries to do baking operations on Athens (004) and the new 005 protocols.
    * Fixes for changes in the format of operations. This is required to properly parse block data in Kiln.
    * Fixes for changes in RPC. This enables Kiln to properly communicate with tezos-node running on new protocol.

  * Support for varying protocols.
    * Kiln now understands that various parameters of the blockchain can change over time due to amendement process.
      This is essential to gather data from blocks of past and current protocols.

  * Bug Fixes
    * Obsidian Public Node APIs
    * UI - Baker alerts now resolve on the tile without a page refresh.

  * Supported networks
    * alphanet, mainnet, zeronet (babylonnet is not supported)

### Known Issues
    * Upgrading Kiln on Ubuntu/Debian overwrites the /etc/kiln/args file

## 0.6.2

  * Update mainnet tezos binaries. This fixes the problem of 100% cpu usage by the tezos-node (and "updater: COMPILATION ERROR" messages). The tezos-node released with v0.6.1 has problem in compiling the new protocol binaries (PsBABY5HQTSk), and this issue surfaced now because mainnet entered the testing phase of the protocol amendement cycle.

## 0.6.1

  * Export logs for Kiln node, baker, and endorser via UI on Ubuntu/Debian.
  * Fix 'Add Baker' flow issues when communicating with tezos-client.
  * Performance improvements and bug fixes for alerts.

tezos-bake-central/common/src/Common/Config.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE ViewPatterns #-}

module Common.Config where

import Control.Applicative (Alternative (..))
import Control.Lens.TH (makeLenses, makePrisms)
import qualified Data.Aeson as Aeson
import Data.Aeson.TH (deriveJSON)
import qualified Data.List.NonEmpty as NE
import qualified Data.Map as Map
import Data.String (IsString)
import qualified Data.Text as T
import Data.Time (NominalDiffTime)
import Data.Version (Version)
import qualified Network.URI.Encode as UriEncode
import Text.Read (readMaybe)
import Text.URI (URI)

import Tezos.Types (ChainId, HashBase58Error (..), NamedChain (..), PublicKeyHash, tryReadPublicKeyHashText)

import Common (defaultTezosCompatJsonOptions)
import Common.URI (Port, mkRootUri)
import ExtraPrelude

changelogUrl :: Text -> Text
changelogUrl branch = "https://gitlab.com/tezos-kiln/kiln/tree/" <> UriEncode.encodeText branch <> "/CHANGELOG.md"

pgConnectionString :: FilePath
pgConnectionString = "pg-connection"

db :: FilePath
db = "db"

route :: FilePath
route = "route"

emailFromAddress :: FilePath
emailFromAddress = "email-from"

chain :: FilePath
chain = "network"

defaultChain :: NamedChain
defaultChain = NamedChain_Mainnet

checkForUpgrade :: FilePath
checkForUpgrade = "check-for-upgrade"

checkForUpgradeDefault :: Bool
checkForUpgradeDefault = True

upgradeBranch :: FilePath
upgradeBranch = "upgrade-branch"

upgradeBranchDefault :: Text
upgradeBranchDefault = "master"

networkGitLabProjectIdDefault :: Text
networkGitLabProjectIdDefault = "3836952"


parseSecondsUnsafe :: Text -> NominalDiffTime
parseSecondsUnsafe = unsafeParse "seconds" $ \a -> case readMaybe (T.unpack a) of
  Nothing -> Left "Not a number of seconds"
  Just b -> Right $ fromIntegral (b :: Integer)

parseBool :: Text -> Bool
parseBool txt
  | v `elem` trues = True
  | v `elem` falses = False
  | otherwise = error $ T.unpack $
      "Can't parse '" <> txt <> "': Expecting one of " <> T.intercalate "/" trues <> " or " <> T.intercalate "/" falses
  where
    trues = ["t", "true", "y", "yes", "on", "enable", "enabled"]
    falses = ["f", "false", "n", "no", "off", "disable", "disabled"]
    v = T.toLower $ T.strip txt

conjList :: Text -> Text -> NE.NonEmpty Text -> Text
conjList comma conj = go
  where
    go xs = case NE.uncons xs of
      (x, Nothing) -> x
      (x, Just (y :| [])) -> x <> conj <> y
      (x, Just xs') -> x <> comma <> go xs'

parseBakerAddr :: Text -> Either Text PublicKeyHash
parseBakerAddr v = do
  unless (T.take 3 v `elem` okPrefixes) $ do
    Left $ (if T.take 3 v == "KT1" then "\"KT1\" addresses cannot bake. Address" else "Baker address") <> " must begin with " <> conjList ", " " or " (NE.map tshow okPrefixes) <> "."
  for_ (T.find (isNothing . flip T.find "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz" . (==)) v) $ \ch ->
    Left $ "The character " <> tshow ch <> " is not allowed in a baker address."
  when (T.length v /= 36) $ Left $ "Baker address is too " <> (if T.length v < 36 then "short" else "long") <> " (must be 36 characters)."
  flip first (tryReadPublicKeyHashText v) $ \case
    HashBase58Error_InvalidPrefix {} -> "This address is outside the valid range for " <> T.take 3 v <> " addresses."
    HashBase58Error_BadChecksum {} -> "This address failed the integrity check. Please check that it has been copied correctly."
    e -> "An unknown error happened, please report this as a bug: " <> tshow e
  where
    okPrefixes :: NE.NonEmpty Text
    okPrefixes = "tz1" :| ["tz2", "tz3"]

parseRootURIUnsafe :: Text -> URI
parseRootURIUnsafe = unsafeParse "URI" mkRootUri

tzscanApiUri :: FilePath
tzscanApiUri = "tzscan-api-uri"

blockscaleApiUri :: FilePath
blockscaleApiUri = "blockscale-api-uri"

nodeConfigFile :: FilePath
nodeConfigFile = "node-config-file"

nodes :: FilePath
nodes = "nodes"

bakers :: FilePath
bakers = "bakers"

parseWithAlias :: (Text -> a) -> Text -> (a, Maybe Text)
parseWithAlias parse txt = case T.breakOn "@" txt of
  (a, T.drop 1 -> b) -> (parse a, if T.null b then Nothing else Just b)

parseNodesUnsafe :: Text -> Map.Map URI (Maybe Text)
parseNodesUnsafe = Map.fromList . parseCommaList (parseWithAlias parseRootURIUnsafe)

parseCommaList :: (Text -> a) -> Text -> [a]
parseCommaList parse = map parse . filter (not . T.null) . map T.strip . T.splitOn ","

parsePublicKeyHashUnsafe :: Text -> PublicKeyHash
parsePublicKeyHashUnsafe = unsafeParse "public key hash" parseBakerAddr

parseBakersUnsafe :: Text -> Map.Map PublicKeyHash (Maybe Text)
parseBakersUnsafe = Map.fromList . parseCommaList (parseWithAlias parsePublicKeyHashUnsafe)

networkGitLabProjectId :: FilePath
networkGitLabProjectId = "network-gitlab-project-id"

tezosReleaseTag :: FilePath
tezosReleaseTag = "tezos-release-tag"

kilnNodeNetPort :: FilePath
kilnNodeNetPort = "kiln-node-net-port"

defaultKilnNodeNetPort :: Port
defaultKilnNodeNetPort = 9733

kilnNodeRpcPort :: FilePath
kilnNodeRpcPort = "kiln-node-rpc-port"

defaultKilnNodeRpcPort :: Port
defaultKilnNodeRpcPort = 8733

kilnDataDir :: FilePath
kilnDataDir = "kiln-data-dir"

defaultKilnDataDir :: FilePath
defaultKilnDataDir = "./.kiln"

kilnNodeCustomArgs :: FilePath
kilnNodeCustomArgs = "kiln-node-custom-args"

kilnBakerCustomArgs :: FilePath
kilnBakerCustomArgs = "kiln-baker-custom-args"

binaryPaths :: FilePath
binaryPaths = "binary-paths"

defaultRightsHistoryWindow :: Int
defaultRightsHistoryWindow = 180

rightsHistoryWindow :: FilePath
rightsHistoryWindow = "rights-history-window"

checkLedgerConnection :: FilePath
checkLedgerConnection = "check-ledger-connection"

defaultCheckLedgerConnection :: Bool
defaultCheckLedgerConnection = True

processRestartMaxDelay :: FilePath
processRestartMaxDelay = "process-restart-max-delay"

defaultProcessRestartMaxDelay :: Int
defaultProcessRestartMaxDelay = 600

resolvedAlertsTtl :: FilePath
resolvedAlertsTtl = "resolved-alerts-ttl"

dalNodeUri :: FilePath
dalNodeUri = "dal-node-uri"

defaultResolvedAlertsTtl :: Int
defaultResolvedAlertsTtl = 4320 -- 3 days

singleQuoted :: (IsString a, Semigroup a) => a -> a
singleQuoted s = "'" <> s <> "'"

unsafeParse :: Text -> (Text -> Either Text a) -> Text -> a
unsafeParse name parse txt = either
  (\msg -> error $ T.unpack $ T.intercalate " "
    ["Invalid"
    , name
    , singleQuoted txt <> ":"
    , msg
    ])
  id
  (parse txt)

parsePortUnsafe :: Text -> Port
parsePortUnsafe = unsafeParse "port number" $ \a -> case readMaybe (T.unpack a) of
  Nothing -> Left "Not a port number"
  Just b -> Right b

newtype UsingNodeOption = UsingCustomNode Aeson.Value
  deriving (Eq, Show, Generic)

instance Aeson.ToJSON UsingNodeOption
instance Aeson.FromJSON UsingNodeOption

data FrontendConfig = FrontendConfig
  { _frontendConfig_chain :: Either NamedChain ChainId
  , _frontendConfig_chainId :: ChainId
  , _frontendConfig_checkForUpgrade :: Bool
  , _frontendConfig_appVersion :: Version
  , _frontendConfig_usingNodeOption :: Maybe UsingNodeOption
  , _frontendConfig_logExportAvailable :: Bool
  , _frontendConfig_ledgerConnectedChecks :: Bool
  , _frontendConfig_tezosGitlabProjectId :: Text
  , _frontendConfig_tezosRelease :: Maybe Text
  } deriving (Eq, Show, Generic, Typeable)

class HasFrontendConfig r where
  frontendConfig :: Lens' r FrontendConfig

instance HasFrontendConfig FrontendConfig where
  frontendConfig = id

combineConfigs :: (Alternative f, Functor g) => f a -> g (f a) -> g (f a)
combineConfigs cfgFromOpts cfgFromFile = fmap ((<|>) cfgFromOpts) cfgFromFile

makePrisms ''UsingNodeOption
makeLenses ''FrontendConfig
concat <$> traverse (deriveJSON $ defaultTezosCompatJsonOptions { Aeson.omitNothingFields = True })
  [ 'FrontendConfig
  ]
